### PR TITLE
Replaced ObjectWaitRequests with unordered map

### PR DESCRIPTION
**Pull request by pschafhalter**
- Replaced `ObjectWaitRequests` with `std::unordered_map<ObjectID, std::vector<WaitRequest *>, UniqueIDHasher>`
- Affects `object_wait_requests_local` and `object_wait_requests_remote` in `PlasmaManagerState`
- Updated all functions involving `ObjectWaitRequests` to use `unordered_map` instead
- Updated comments

### CODE REVIEW

File: src/plasma/plasma_manager.cc [hunk 1]
```diff
@@ -344,9 +337,9 @@ void object_table_subscribe_callback(ObjectID object_id,
                                      const char *manager_vector[],
                                      void *context);
 
-ObjectWaitRequests **object_wait_requests_table_ptr_from_type(
-    PlasmaManagerState *manager_state,
-    int type) {
+std::unordered_map<ObjectID, std::vector<WaitRequest *>, UniqueIDHasher>
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: I think you generally don't want to return pointers to STL containers, you should just return the container itself.

I *think* that in this case we want to return a 

```c++
std::unordered_map<ObjectID, std::vector<WaitRequest *>, UniqueIDHasher> &
```

but it'd be good if @pcmoritz could comment as well.

When we use it, it should also be a reference. If you do

```c++
auto x = object_wait_requests_ptr_from_type(manager_state, type);
```

then will `x` be a reference or not?

Comment (pschafhalter) [reply src/plasma/plasma_manager.cc]: I think you're right, that should work. I'll implement that fix.

Comment (pcmoritz) [reply src/plasma/plasma_manager.cc]: Yeah, a reference here is better; on the caller side I have a feeling that
```
auto& x = object_wait_requests_ptr_from_type(manager_state, type);
```
is the way to go, but I'm not sure if that works.

**Comment by pschafhalter** [comment]
Looks like there's a memory leak. I'll try to find the issue.

**Comment by robertnishihara** [comment]
Are you referring to the failed valgrind test? If so, that test has been failing consistently for a little while now and so that is probably unrelated to this PR.

**Comment by pschafhalter** [comment]
Yea, I was referring to that. If that's unrelated to the PR, then it should be ready to merge -- I updated the pointers to references.

**Comment by robertnishihara** [comment]
Looks good to me. I'll merge it once the tests run.

**Pull request merged**